### PR TITLE
vo_gpu_next: add dynamic hook parameters

### PR DESCRIPTION
**Pull request by haasn**
TODO:
- add more options to map?
- document them somewhere
- testing

Fixes: https://github.com/haasn/libplacebo/issues/230
Fixes: https://github.com/mpv-player/mpv/issues/8206

Read this before you submit this pull request:
https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md

Reading this link and following the rules will get your pull request reviewed
and merged faster. Nobody wants lazy pull requests.

### CODE REVIEW

**Comment by haasn** [comment]
This adds a special helper inside `vo_gpu_next` to map properties of the `mp_image` directly. It is not strictly speaking what was requested in #8206, since that was asking for the ability to map *arbitrary* mpv parameters. That seems much harder to do, however.

Maybe this design is not the best, and it would be better to instead handle this at the mpv options layer, by transparently forwarding mpv properties to the shader opts dictionary.

File: video/out/vo_gpu_next.c [hunk 1]
```diff
@@ -1862,6 +1869,28 @@ static void update_lut(struct priv *p, struct user_lut *lut)
...
+    for (int i = 0; i < hook->num_parameters; i++) {
+        const struct pl_hook_par *hp = &hook->parameters[i];
+        for (int n = 0; n < MP_ARRAY_SIZE(opts); n++) {
+            if (strcmp(hp->name, opts[n].name) != 0)
+                continue;
+
+            switch (hp->type) {
+                case PL_VAR_FLOAT: hp->data->f = opts[n].value; break;
+                case PL_VAR_SINT:  hp->data->i = round(opts[n].value); break;
+                case PL_VAR_UINT:  hp->data->u = round(opts[n].value); break;
```
Comment (kasper93) [inline video/out/vo_gpu_next.c:1888]: It is UB to assign values form outside the variable range. Should be clampled here, also consider using `lrintf` that would help.

File: video/out/vo_gpu_next.c [hunk 2]
```diff
@@ -865,6 +865,9 @@ static void update_tm_viz(struct pl_color_map_params *params,
     params->visualize_hue = M_PI / 4.0;
 }
 
+static void update_hook_opts_dynamic(struct priv *p, const struct pl_hook *hook,
+                                     const struct mp_image *mpi);
+
```
Comment (kasper93) [inline video/out/vo_gpu_next.c:870]: Is this forward deceleration necessary?

**Comment by Cqoicebordel** [comment]
Not sure if I should comment here, please excuse me if not.

I'm looking at rendering spherical videos using un360 (https://gist.github.com/tesu/196db5421559de3e9555d4f9da9d847d). There are everything in there, except navigation in the sphere.  
The approach I feel would fit, and fits this PR, is to be able to use newly created variables : we use input.conf to define values when keys are pressed (like KP_8 for up), and modify named variable (`pitch = pitch + 0.1`). Then, the glsl is hooked to those same names, and use them dynamically.

Another approach would be to call functions of the glsl when we press a key, but that doesn't fit this PR ;)

**Comment by kasper93** [comment]
Superseded by #15239

**Pull request closed without merging**